src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,373 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.InlineMethod
+{
+    internal abstract partial class AbstractInlineMethodRefactoringProvider<
+            TInvocationSyntax,
+            TExpressionSyntax,
+            TMethodDeclarationSyntax,
+            TStatementSyntax>
+        : CodeRefactoringProvider
+        where TExpressionSyntax : SyntaxNode
+        where TInvocationSyntax : TExpressionSyntax
+        where TMethodDeclarationSyntax : SyntaxNode
+        where TStatementSyntax : SyntaxNode
+    {
+        private readonly ISyntaxFacts _syntaxFacts;
+        private readonly ISemanticFactsService _semanticFactsService;
+
+        protected abstract TExpressionSyntax? GetInlineExpression(TMethodDeclarationSyntax calleeMethodDeclarationSyntaxNode);
+        protected abstract SyntaxNode? GetEnclosingMethodLikeNode(SyntaxNode syntaxNode);
+        protected abstract SyntaxNode GenerateTypeSyntax(ITypeSymbol symbol, bool allowVar);
+        protected abstract TExpressionSyntax Parenthesize(TExpressionSyntax expressionNode);
+        protected abstract bool TryGetInlineSyntaxNodeAndReplacementNodeForDelegate(
+            TInvocationSyntax calleeInvocationNode,
+            IMethodSymbol calleeMethodSymbol,
+            TExpressionSyntax inlineExpressionNode,
+            TStatementSyntax statementContainsCallee,
+            SyntaxGenerator syntaxGenerator,
+            out SyntaxNode? inlineSyntaxNode,
+            out SyntaxNode? syntaxNodeToReplace);
+
+        /// <summary>
+        /// Check if <paramref name="expressionNode"/> could be used as an Expression in ExpressionStatement
+        /// </summary>
+        protected abstract bool IsValidExpressionUnderStatementExpression(TExpressionSyntax expressionNode);
+
+        /// <summary>
+        /// A preferred name used to generated a declaration when the
+        /// inline method has a return value but is not assigned to a variable.
+        /// Example:
+        /// void Caller()
+        /// {
+        ///     Callee();
+        /// }
+        /// int Callee()
+        /// {
+        ///     return 1;
```

Comment:
def an interesting example.  it's not clear to me why iw ant a temp here.  Let's go with something more intersting:

```
        /// void Caller()
        /// {
        ///     DoWork();
        /// }
        /// int Callee()
        /// {
        ///     return SomeWork();
        ///  }
```

Should i end with:

```int temp = SomeWork();``` or just ```SomeWork();```.  I think i just want the latter.

Reply:
the only time i'd want the former would be if the expression being inlined was not a legal statement-expression.

Reply:
Yeah, that's what I mean, will update the comments